frontend/src/metabase/query_builder/components/DatasetEditor/DatasetEditorInner.tsx

```diff
@@ -554,10 +558,13 @@ const _DatasetEditorInner = (props: DatasetEditorInnerProps) => {
     const canBeDataset = checkCanBeModel(question);
     const isBrandNewDataset = !question.id();
     let questionWithUpdatedSettings = question;
-    if (
-      !!tempModelSettings.display &&
-      tempModelSettings.display !== question.display()
-    ) {
+    /**
+     * When updating 'display' setting, we need to check if there were actually
+     * any user-driven changes, because the same temp setting value is used
+     * to display proper UI when switching between 'Columns' and 'Settings' tabs.
+     * (see comment for `tempModelSettings`)
+     */
```

Comment:
This logic is hard to follow. I'm not confident it works as expected.

----

:x: 1. Unnecessary unsaved changes warning when creating a new model

https://github.com/user-attachments/assets/85e9f81d-95c3-4c41-8d91-95dc75eec48e

----

:x: 2. Unnecessary unsaved changes warning when editing an existing model

https://github.com/user-attachments/assets/bcb80584-9634-4a4c-8403-b69f2458f755

----

Let's add repros for these cases and test it more :pray:

Reply:
Yeah, that's a known issue, which was actually always there. I'll try to take another look, because it may involve changing some very old logic for changes detection, and I was trying to avoid that.